[torchgen] Move Executorch custom ops logic into torchgen

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89596
* #90806
* __->__ #90099
* #90098

## Logic to handle custom ops
We generate files for custom ops, so that they can be registered into PyTorch.

Generated files:
* `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.
* `Register{dispatch_key}Stub.cpp` (dispatch_key = CPU), register placeholder kernels for custom ops. Only used when there's no custom op kernel available.

As an example:
```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

larryliu0820 committed:
[torchgen] Move Executorch custom ops logic into torchgen



Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597382).

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops

We generate files for custom ops, so that they can be registered into PyTorch.

Generated file is `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.

As an example: 

```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops

We generate files for custom ops, so that they can be registered into PyTorch.

Generated file is `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.

As an example: 

```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops

We generate files for custom ops, so that they can be registered into PyTorch.

Generated file is `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.

As an example: 

```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops

We generate files for custom ops, so that they can be registered into PyTorch.

Generated file is `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.

As an example: 

```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops

We generate files for custom ops, so that they can be registered into PyTorch.

Generated file is `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.

As an example: 

```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597382).

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops

We generate files for custom ops, so that they can be registered into PyTorch.

Generated file is `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.

As an example: 

```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597382).

ezyang commented on tools/test/test_executorch_custom_ops.py:
```diff
@@ -0,0 +1,74 @@
+        expected = """
+at::Tensor wrapper_Tensor_foo(const at::Tensor & self) {
+    return self;
+}
```
This tests are very awkward. At the very least, can we use an expect test for this?

larryliu0820 replied:
Any code pointer for existing expect tests?

ezyang replied:
search assertExpectedInline

ezyang commented on torchgen/executorch/api/custom_ops.py:
```diff
@@ -0,0 +1,163 @@
+    """
+
+
+def gen_custom_ops(
```
Hmm, this doesn't follow the convention for the other api files, which is that the actual file manager heavy lifting happens at the top level executorch driver script (wherever that lives). The main reason for this is circular deps, as you have noticed when doing the imports for this file. Consider reorganizing it in that way.

ezyang commented on torchgen/executorch/api/custom_ops.py:
```diff
@@ -0,0 +1,163 @@
+                f" Got outs {str(f.func.arguments.out)} but returns {str(f.func.returns)}"
+            )
+            # returns a tuple of out arguments
+            tensor_type = "at::Tensor &"
```
Is there any reason for the executorch generated C++ code to return mutable Tensor references? I can't think of any good reason. It's not like you're using them right? You could probably return void and it would be all fine.

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops
We generate files for custom ops, so that they can be registered into PyTorch.

Generated files:
* `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.
* `Register{dispatch_key}Stub.cpp` (dispatch_key = CPU), register placeholder kernels for custom ops. Only used when there's no custom op kernel available.

As an example:
```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops
We generate files for custom ops, so that they can be registered into PyTorch.

Generated files:
* `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.
* `Register{dispatch_key}Stub.cpp` (dispatch_key = CPU), register placeholder kernels for custom ops. Only used when there's no custom op kernel available.

As an example:
```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41597382/)!

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41597382).

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42037214).

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops
We generate files for custom ops, so that they can be registered into PyTorch.

Generated files:
* `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.
* `Register{dispatch_key}Stub.cpp` (dispatch_key = CPU), register placeholder kernels for custom ops. Only used when there's no custom op kernel available.

As an example:
```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```

Differential Revision: [D41597382](https://our.internmc.facebook.com/intern/diff/D41597382/)


[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Move Executorch custom ops logic into torchgen"


## Logic to handle custom ops
We generate files for custom ops, so that they can be registered into PyTorch.

Generated files:
* `Register{dispatch_key}CustomOps.cpp` (dispatch_key = CPU), it's basically the same as vanilla PyTorch `RegisterCPU.cpp`. The only difference is that we bind to native functions directly.
* `Register{dispatch_key}Stub.cpp` (dispatch_key = CPU), register placeholder kernels for custom ops. Only used when there's no custom op kernel available.

As an example:
```cpp
namespace {

at::Tensor & wrapper_out_unsqueeze_out(const at::Tensor & self, int64_t dim, at::Tensor & out) {
    // No device check


  // DeviceGuard omitted
  return torch::executor::native::unsqueeze_out(self, dim, out);
}
} // anonymous namespace

TORCH_LIBRARY_IMPL(aten, CPU, m) {

m.impl("unsqueeze.out",
TORCH_FN(wrapper_out_unsqueeze_out));
}
```


[ghstack-poisoned]
